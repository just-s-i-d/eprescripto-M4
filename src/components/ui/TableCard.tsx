import { SetStateAction, useEffect, useState } from "react";
import { Input, Table } from "antd";

import { SearchOutlined } from "@ant-design/icons";

import ErrorBoundary from "@components/ErrorBoundary";
import { ColumnsType } from "antd/es/table";

type TableCardPropType = {
  tableData?: Record<string, string | number>[];
  columns: ColumnsType<Record<string, string | number>>;
  setLoading: React.Dispatch<SetStateAction<boolean>>;
  loading: boolean;
  pageSize?: number;
  className?: string;
  setRefresh: React.Dispatch<SetStateAction<boolean>>;
  error: boolean;
};
const TableCard = ({
  tableData,
  columns,
  pageSize = 10,
  className,
  error,
  setLoading,
  setRefresh,
  loading,
}: TableCardPropType) => {
  const [filteredData, setFilteredData] = useState(tableData);
  const handleSearch = (value: string) => {
    setLoading(true);
    const newFilteredData = tableData?.filter((item) =>
      Object.values(item).some(
        (field) =>
          field && field.toString().toLowerCase().includes(value.toLowerCase()),
      ),
    );
    setFilteredData(newFilteredData);
    setLoading(false);
  };
  useEffect(() => {
    setFilteredData(tableData);
  }, [tableData]);
  return (
    <div className="marginTop-12 minHeight-74">
      <div className="margin-bottom-12">
        <Input
          placeholder="Search"
          className="custom-search"
          onChange={(event) => handleSearch(event.target.value)}
          prefix={<SearchOutlined />}
        />
      </div>
      <ErrorBoundary
        error={error}
        refreshComponent={() => setRefresh((prev) => !prev)}
      >
        <Table
          data-testid="custom-table"
          className={`custom-table ${className || ""}`}
          columns={columns}
          dataSource={filteredData}
          loading={loading}
          pagination={{ pageSize: pageSize }}
        />
      </ErrorBoundary>
    </div>
  );
};

export default TableCard;
