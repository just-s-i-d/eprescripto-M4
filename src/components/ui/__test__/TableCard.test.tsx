import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import TableCard from "../TableCard";
import "@test/matchMedia";

const mockTableData = [
  { id: 1, name: "John Doe", age: 30 },
  { id: 2, name: "Jane Doe", age: 25 },
];

const mockColumns = [
  { title: "ID", dataIndex: "id", key: "id" },
  { title: "Name", dataIndex: "name", key: "name" },
  { title: "Age", dataIndex: "age", key: "age" },
];

describe("TableCard Component", () => {
  it("Snapshot test", () => {
    const { asFragment } = render(
      <TableCard
        tableData={mockTableData}
        columns={mockColumns}
        setLoading={jest.fn()}
        setRefresh={jest.fn()}
        loading={false}
        error={false}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders TableCard with search input and table", () => {
    render(
      <TableCard
        tableData={mockTableData}
        columns={mockColumns}
        setLoading={jest.fn()}
        setRefresh={jest.fn()}
        loading={false}
        error={false}
      />,
    );
    expect(screen.getByPlaceholderText("Search")).toBeInTheDocument();
    expect(screen.getByTestId("custom-table")).toBeInTheDocument();
  });
  it("Error handle test", () => {
    const mockedSetRefresh = jest.fn();
    render(
      <TableCard
        tableData={[]}
        columns={mockColumns}
        setLoading={jest.fn()}
        setRefresh={mockedSetRefresh}
        loading={false}
        error={true}
      />,
    );
    expect(screen.getByText("Something went wrong")).toBeInTheDocument();
    const tryAgain = screen.getByText("Try again");
    expect(tryAgain).toBeInTheDocument();
    fireEvent.click(tryAgain);
    expect(mockedSetRefresh).toHaveBeenCalled();
  });
  it("shows no data state in the table", () => {
    render(
      <TableCard
        tableData={[]}
        columns={mockColumns}
        setLoading={jest.fn()}
        setRefresh={jest.fn()}
        loading={false}
        error={false}
      />,
    );
    const noData = screen.getByText("No data");
    expect(screen.getByTestId("custom-table")).toBeInTheDocument();
    expect(noData).toHaveClass("ant-empty-description");
  });

  it("filters table data when search input changes", async () => {
    render(
      <TableCard
        tableData={mockTableData}
        columns={mockColumns}
        setLoading={jest.fn()}
        setRefresh={jest.fn()}
        loading={false}
        error={false}
      />,
    );

    fireEvent.change(screen.getByPlaceholderText("Search"), {
      target: { value: "John" },
    });

    await waitFor(() => {
      expect(screen.getByText("John Doe")).toBeInTheDocument();
      expect(screen.queryByText("Jane Doe")).toBeNull();
    });
  });
});
