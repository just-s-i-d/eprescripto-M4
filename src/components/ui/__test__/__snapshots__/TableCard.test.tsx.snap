// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCard Component Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="marginTop-12 minHeight-74"
  >
    <div
      class="margin-bottom-12"
    >
      <span
        class="ant-input-affix-wrapper custom-search css-dev-only-do-not-override-gzal6t css-dev-only-do-not-override-gzal6t"
      >
        <span
          class="ant-input-prefix"
        >
          <span
            aria-label="search"
            class="anticon anticon-search"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </span>
        </span>
        <input
          class="ant-input css-dev-only-do-not-override-gzal6t"
          placeholder="Search"
          type="text"
          value=""
        />
      </span>
    </div>
    <div
      class="ant-table-wrapper custom-table  css-dev-only-do-not-override-gzal6t"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-gzal6t"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table css-dev-only-do-not-override-gzal6t"
            data-testid="custom-table"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        ID
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Name
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Age
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        1
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        John Doe
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        30
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        2
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        Jane Doe
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        25
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
          <ul
            class="ant-pagination ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-gzal6t"
          >
            <li
              aria-disabled="true"
              class="ant-pagination-prev ant-pagination-disabled"
              title="Previous Page"
            >
              <button
                class="ant-pagination-item-link"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  aria-label="left"
                  class="anticon anticon-left"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="left"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                    />
                  </svg>
                </span>
              </button>
            </li>
            <li
              class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
              tabindex="0"
              title="1"
            >
              <a
                rel="nofollow"
              >
                1
              </a>
            </li>
            <li
              aria-disabled="true"
              class="ant-pagination-next ant-pagination-disabled"
              title="Next Page"
            >
              <button
                class="ant-pagination-item-link"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  aria-label="right"
                  class="anticon anticon-right"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="right"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                    />
                  </svg>
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
